core/dom/range.js

```diff
@@ -2818,6 +2818,69 @@ CKEDITOR.dom.range = function( root ) {
 			reference.remove();
 		},
 
+		/**
+		 * Returns Array of rects from dom elements contained within current ranges.
+		 *
+		 * @since 4.9.0
+		 * @private
+		 */
+		getClientRects: function() {
+			// Extending empty object wth rect, to prevent inheriting from DOMRect, same approach as in CKEDITOR.dom.element.getClientRect().
+			function convertRect( rect ) {
+				var newRect = CKEDITOR.tools.extend( {}, rect );
+				// Some browsers might not return width and height.
+				!newRect.width && ( newRect.width = newRect.right - newRect.left );
+				!newRect.height && ( newRect.height = newRect.bottom - newRect.top );
+				return newRect;
+			}
+
+			// Fallback helper for browsers that don't support native getClientRects()
+			function getRect( context ) {
+				var bookmark = context.createBookmark(),
+					start = bookmark.startNode,
+					end = bookmark.endNode,
+					rects,
+					rect = {};
+
+				start.setText = '';
+				start.removeStyle( 'display' );
+
+				if ( end ) {
+					end.setText = '';
+					end.removeStyle( 'display' );
+					rects = [ start.getClientRect(), end.getClientRect() ];
+					start.remove();
+					end.remove();
+				} else {
+					rects = [ start.getClientRect(), start.getClientRect() ];
+					start.remove();
+				}
+				rect.right = Math.max( rects[ 0].right, rects[ 1 ].right );
+				rect.bottom = Math.max( rects[ 0].bottom, rects[ 1 ].bottom );
+				rect.left = Math.min( rects[ 0 ].left, rects[ 1 ].left );
+				rect.top = Math.min( rects[ 0 ].top, rects[ 1 ].top );
+				rect.width = Math.abs( rects[ 0 ].left - rects[ 1 ].left );
+				rect.height = rect.bottom - rect.top;
+
+				return rect;
+			}
+
+			if ( this.document.$.getSelection !== undefined ) {
+				var sel = this.document.$.getSelection(),
```

Comment:
You're working on a global selection here, which is not the same as your context selection.

Just imagine that you create a new range _in memory_ in order to use it later or whatever, but you do no actually select it. E.g.

```js
var rng = new CKEDITOR.dom.range( CKEDITOR.document )
rng.setStart( … );
rng.setEnd( … );
var rects = rng.getClientRects();
```
👆 this case needs to be included in unit tests.

Reply:
I can see also a design flaw in our Range implementation which will make your task harder - AFAICT we're not keeping a reference to the native Range object.

If that's the case your only solution for this will be to temporarily create a new range, use it to _probe_ what you need to check and then destroy it.

Reply:
Also there is a need to handle our _fake selection_ - however we could extract it into a separate ticket. But please familiarize your self with this aspect already (talk to other devs if you need a help with that).